### PR TITLE
Support newer Jakarta Faces API versions (3.0, 4.0, 4.1)

### LINKED ISSUE #8: Support newer Jakarta Faces API versions (3.0, 4.0, 4.1)
Would be great to get support for the newer Jakarta Faces API versions 3.0, 4.0 and 4.1 (see https://jakarta.ee/specifications/faces/). For now Eclipse seems to be stuck at version 2.3.

Related bugzilla reports:
- https://bugs.eclipse.org/bugs/show_bug.cgi?id=581893
- https://bugs.eclipse.org/bugs/show_bug.cgi?id=577596

### PR DESCRIPTION
**Pull request by bvfalcon**
Closes #8 

Solution details:
1) New Facets (for Faceted Project) added: 3.0, 4.0, 4.1
2) New presets "Jakarta Faces 4.0", "Jakarta Faces 4.1", "JavaServer Faces 2.3" added, presets for outdated JSF versions removed
3) Added new menu item "New -> JSF File" with own wizard instead of creating JSF files with JSP files wizard
4) Added support for jakarta taglib with new namespaces

Known issues:
1) Documentation is not updated
2) Unit- and Integration-tests are not actualized, only adapted for code changes

### CODE REVIEW

**Comment by urbandroid** [comment]
needs attention

**Comment by erik-brangs** [comment]
Removing support for JSF 2.2 and earlier removes support for JSF for Java EE 7 and earlier. Is this really necessary? AFAIK Support for other Jakarta specifications (e.g. JPA) kept compatibility with earlier EE versions.

The downloadable libraries are kept at the website repository at https://gitlab.eclipse.org/eclipse/webtools/webtools-website . They're under ``jsf/jsf-library``. You can just submit a pull request there.

**Comment by nitind** [comment]
> Removing support for JSF 2.2 and earlier removes support for JSF for Java EE 7 and earlier. Is this really necessary?

It's not the sort of thing I'd approve for merging.

Most of the functionality for existing versions should be able to be adapted to work off of one set of packages or the other based on the deployment descriptor contents (if present), the project's facet version (if present), or by brute-force checking the project's Java Build Path (e.g. https://github.com/eclipse-sourceediting/sourceediting/blob/bdf63c0f061b10059ffcec9350ec8da210b512f4/web/bundles/org.eclipse.jst.jsp.core/src/org/eclipse/jst/jsp/core/internal/contenttype/DeploymentDescriptorPropertyCache.java#L1049 and the way it spiders out to do all of that to pick a Servlet API version). I don't undervalue how much effort that entails.

**Comment by bvfalcon** [comment]
> The downloadable libraries are kept at the website repository at https://gitlab.eclipse.org/eclipse/webtools/webtools-website . They're under `jsf/jsf-library`. You can just submit a pull request there.

Thanks a lot, [MR](https://gitlab.eclipse.org/eclipse/webtools/webtools-website/-/merge_requests/6) is created.

File: jsf/plugins/org.eclipse.jst.jsf.common.ui/src/org/eclipse/jst/jsf/common/ui/IFileFolderConstants.java [hunk 1]
```diff
@@ -45,12 +45,17 @@ public interface IFileFolderConstants {
 	 */
 	public static final String EXT_JAVA = "java"; //$NON-NLS-1$
 
-
 	/**
 	 * jsp file extension
+	 * @deprecated : JSP is not a part of JSF
```
Comment (erik-brangs) [inline jsf/plugins/org.eclipse.jst.jsf.common.ui/src/org/eclipse/jst/jsf/common/ui/IFileFolderConstants.java:50]: I think this would profit from a more thorough explanation. The comment should make it clear that newer versions of Jakarta Faces no longer support JSP. At the very least, it should say something like "Support for JSP is no longer available since Jakarta Faces 4.0".

The ``@deprecated`` should be removed. JSP is still valid for older versions of the specification.

File: jsf/plugins/org.eclipse.jst.jsf.facelet.core/src/org/eclipse/jst/jsf/facelet/core/internal/registry/taglib/faceletTaglib/FaceletTaglibVersion.java [hunk 2]
```diff
@@ -76,7 +76,34 @@ public enum FaceletTaglibVersion implements Enumerator
 	 * @generated
 	 * @ordered
 	 */
-	_23(3, "_23", "2.3");  //$NON-NLS-1$//$NON-NLS-2$
```
Comment (erik-brangs) [inline jsf/plugins/org.eclipse.jst.jsf.facelet.core/src/org/eclipse/jst/jsf/facelet/core/internal/registry/taglib/faceletTaglib/FaceletTaglibVersion.java:79]: This is generated code. The ecore files from the model directory need to be updated. See https://eclipsesource.com/blogs/tutorials/emf-tutorial/ for an introduction to EMF (Eclipse Modelling Framework).

File: jsf/plugins/org.eclipse.jst.jsf.facesconfig.ui/src/org/eclipse/jst/jsf/facesconfig/ui/pageflow/util/JSPUtil.java [hunk 3]
```diff
@@ -25,6 +25,8 @@
  * utility class for JSP related information.
  * 
  * @author Yang Liu
+ * @deprecated
```
Comment (erik-brangs) [inline jsf/plugins/org.eclipse.jst.jsf.facesconfig.ui/src/org/eclipse/jst/jsf/facesconfig/ui/pageflow/util/JSPUtil.java:28]: Remove the ``@deprecated`` tag.

File: jsf/plugins/org.eclipse.jst.jsf.ui/META-INF/MANIFEST.MF [hunk 4]
```diff
@@ -20,12 +20,14 @@ Require-Bundle: org.eclipse.jst.jsf.core;bundle-version="[1.1.0,2.0.0)",
...
  org.eclipse.emf.ecore;bundle-version="[2.2.0,3.0.0)",
- org.eclipse.jst.jsp.ui;bundle-version="[1.3.0,2.0.0)",
+ org.eclipse.jst.jsf.ui;bundle-version="[1.3.0,2.0.0)",
  org.eclipse.ui.forms;bundle-version="[3.2.0,4.0.0)",
  org.eclipse.wst.sse.core;bundle-version="[1.0.101,2.0.0)",
  org.eclipse.ui.forms;bundle-version="[3.2.0,4.0.0)",
  org.eclipse.jst.jsf.common.runtime;bundle-version="[1.0.0,2.0.0)",
+ org.eclipse.jst.jsf.core;bundle-version="[1.1.0,2.0.0)",
  org.eclipse.jst.jsp.core;bundle-version="[1.1.0,2.0.0)",
+ org.eclipse.jst.jsp.ui;bundle-version="[1.1.0,2.0.0)",
```
Comment (erik-brangs) [inline jsf/plugins/org.eclipse.jst.jsf.ui/META-INF/MANIFEST.MF:30]: Is this actually correct? The range previously was ``"[1.3.0,2.0.0)"``.

The diff would be more readable if you just kept the old line for jsp at its original place and added the jsf one here.

File: jsf/plugins/org.eclipse.jst.pagedesigner.jsf.ui/src/org/eclipse/jst/pagedesigner/jsf/ui/sections/JSFCoreConvertDateTimeGroup.java [hunk 5]
```diff
@@ -179,7 +180,7 @@ private void updateFieldStatus()
         String dateStyle = _dateStyleField.getText();
```
Comment (erik-brangs) [inline jsf/plugins/org.eclipse.jst.pagedesigner.jsf.ui/src/org/eclipse/jst/pagedesigner/jsf/ui/sections/JSFCoreConvertDateTimeGroup.java:180]: The type checks are hard to read. How about some aptly named helper methods?

File: jsf/plugins/org.eclipse.jst.pagedesigner/src/org/eclipse/jst/pagedesigner/converter/AbstractTagConverter.java [hunk 6]
```diff
@@ -362,7 +362,7 @@ protected String mapValue(String value) {
 		} else {
 			// preview mode. let's try to display the value.
 			try {
-				//Bug 319317 - Third-party plug-in providing javax.servlet.jsp.el version 2.1 or greater breaks WPE preview
+				//Bug 319317 - Third-party plug-in providing jakarta.el version 2.1 or greater breaks WPE preview
```
Comment (erik-brangs) [inline jsf/plugins/org.eclipse.jst.pagedesigner/src/org/eclipse/jst/pagedesigner/converter/AbstractTagConverter.java]: This refers to the title of the issue, https://bugs.eclipse.org/bugs/show_bug.cgi?id=319317 .

AFAIK at least a copy of the bugzilla issues will be kept. Therefore, this comment will probably still useful to find the original issue and should not be changed.

File: jsf/plugins/org.eclipse.jst.pagedesigner/src/org/eclipse/jst/pagedesigner/preview/PageExpressionContext.java [hunk 7]
```diff
@@ -112,7 +112,7 @@ public void popPageVarProvider(IPageVariablesProvider provider) {
 	 */
 	public Object evaluateExpression(String expression, Class expectedClass, Map options)
 			throws ELException {
-		//Bug 319317 - Third-party plug-in providing javax.servlet.jsp.el version 2.1 or greater breaks WPE preview
+		//Bug 319317 - Third-party plug-in providing jakarta.servlet.jsp.el version 2.1 or greater breaks WPE preview
```
Comment (erik-brangs) [inline jsf/plugins/org.eclipse.jst.pagedesigner/src/org/eclipse/jst/pagedesigner/preview/PageExpressionContext.java]: Revert this change. The `javax` is part of the title of the issue.

File: jsf/tests/org.eclipse.jst.jsf.core.tests/src/org/eclipse/jst/jsf/core/tests/facet/JsfLibraryValidatorTest.java [hunk 8]
```diff
@@ -14,7 +14,6 @@
 
 package org.eclipse.jst.jsf.core.tests.facet;
 
-
```
Comment (erik-brangs) [inline jsf/tests/org.eclipse.jst.jsf.core.tests/src/org/eclipse/jst/jsf/core/tests/facet/JsfLibraryValidatorTest.java:17]: Remove this whitespace change or put it into a separate commit.

**Review by erik-brangs — CHANGES_REQUESTED** [review]
I have not flagged the parts where both Java EE and Jakarta EE need to be supported. In general, support for old versions of Java EE should be kept.

I'm not sure how the correct approach would look. The important part is to keep compatibility, both for existing clients and users who still need support for pre-Jakarta-JSF.

My personal opinion is that you should split changes that are not directly related (e.g. the JSFUIPlugin rename) into separate commits. They can be squashed later if needed.

You also need to update the copyright headers for the files you change before this can be merged. The year needs to be updated at least.

File: jsf/plugins/org.eclipse.jst.pagedesigner/src/org/eclipse/jst/pagedesigner/preview/PageExpressionContext.java (hunk 7 above)
Comment (bvfalcon) [reply jsf/plugins/org.eclipse.jst.pagedesigner/src/org/eclipse/jst/pagedesigner/preview/PageExpressionContext.java]: change is [reverted](https://github.com/eclipse-jsf/webtools.jsf/pull/9/commits/4b40446e621ee49ab4464eb0eee9e399dd144f9c).

File: jsf/plugins/org.eclipse.jst.pagedesigner/src/org/eclipse/jst/pagedesigner/converter/AbstractTagConverter.java (hunk 6 above)
Comment (bvfalcon) [reply jsf/plugins/org.eclipse.jst.pagedesigner/src/org/eclipse/jst/pagedesigner/converter/AbstractTagConverter.java]: change is [reverted](https://github.com/eclipse-jsf/webtools.jsf/pull/9/commits/61c1cb00662c35bde51b31b272c8b43b539ac778).

**Review by urbandroid — APPROVED** [review]
looks good to me

**Comment by bvfalcon** [comment]
Close this PR: changes will be done with another PR considering notices.

[MR](https://gitlab.eclipse.org/eclipse/webtools/webtools-website/-/merge_requests/6) is still necessary

**Comment by nitind** [comment]
https://gitlab.eclipse.org/eclipse/webtools/webtools-website/-/merge_requests/6 merged. GitLab's not great at notifications.
